corehq/form_processor/backends/sql/update_strategy.py

```diff
@@ -302,3 +349,56 @@ def _apply_form_transaction(self, transaction):
             filtered_updates = [u for u in case_updates if u.id == self.case.case_id]
             for case_update in filtered_updates:
                 self._apply_case_update(case_update, form)
+
+
+def _transaction_sort_key_function(case):
+    xform_ids = list(case.xform_ids)
+    fudge_factor = datetime.timedelta(hours=12)
+
+    def cc_cmp(first, second):
+        if first > second:
+            return 1
+        if first < second:
+            return -1
+        return 0
+
+    def _transaction_cmp(first_transaction, second_transaction):
+        # if the forms aren't submitted by the same user, just default to server dates
+        if first_transaction.user_id != second_transaction.user_id:
+            return cc_cmp(first_transaction.server_date, second_transaction.server_date)
+
+        if first_transaction.form_id and first_transaction.form_id == second_transaction.form_id:
+            # short circuit if they are from the same form
+            return cc_cmp(
+                _type_sort(first_transaction.type),
+                _type_sort(second_transaction.type)
+            )
+
+        if not (first_transaction.server_date and second_transaction.server_date):
+            raise MissingServerDate()
+
+        # checks if the dates received are within a particular range
+        if abs(first_transaction.server_date - second_transaction.server_date) > fudge_factor:
+            return cc_cmp(first_transaction.server_date, second_transaction.server_date)
+
+        def _sortkey(transaction):
+            def form_cmp(form_id):
+                return xform_ids.index(form_id) if form_id in xform_ids else sys.maxsize
```

Comment:
both `xform_ids.index(...)` and `... in xform_ids` are O(n).
```suggestion
                try:
                    return xform_ids.index(form_id)
                except ValueError:
                    return sys.maxsize
```

Reply:
3fabacf